lib/utils.nix

```
@@ -125,27 +125,185 @@ rec {
     end
   '';
 
+  /**
+    Generic "heredoc-style" escape utility.
+
+    # Args
+
+    - string: The string to escape.
+    - buildSurround: A function that maps Integer to Surround. Described in detail below.
+    - surroundWithNewlines: A boolean that impacts how to surround the string when escaping it.
+
+    A Surround is an attr set that looks like this:
+
+    ```nix
+    {
+      open = "(";
+      close = ")";
+    }
+    ```
+
+    `buildSurround` is a function that takes a depth Integer and returns a
+    Surround. For exmple, to build the previous Surround:
+
+    ```nix
+    nix-repl> buildSurround = depth: {
+      open = lib.strings.replicate depth "(";
+      close = lib.strings.replicate depth ")";
+    }
+    nix-repl> buildSurround 1
+    {
+      close = ")";
+      open = "(";
+    }
+    ```
+
+    We invoke `buildSurround` builder starting a depth 0, and increasing the
+    depth until we have a Surround where `surround.close` is not in the
+    string we're escaping.
+
+    We return the original string, first (possibly, depending on
+    `surroundWithNewlines`) surrounded by newlines , and secondly surrounded by
+    the Surround we found above.
+    ```
+  */
+  escapeString =
+    {
+      string,
+      buildSurround,
+      surroundWithNewlines,
+    }:
+    let
+      findSurround =
+        {
+          string,
+          buildSurround,
+          depth,
+        }:
```

If we keep things as they are, there is no need to pass `string` and `buildSurrond` into `findSurround`, because they are already lexically in scope (i.e. in scope via closure):

```suggestion
      findSurround =
        depth:
```